django/db/models/functions/comparison.py

```diff
@@ -112,6 +113,38 @@ def as_sqlite(self, compiler, connection, **extra_context):
         return super().as_sqlite(compiler, connection, function='MAX', **extra_context)
 
 
+class JSONObject(Func):
+    function = 'JSON_OBJECT'
+    output_field = JSONField()
+
+    def __init__(self, **fields):
+        expressions = []
+        for key, value in fields.items():
+            expressions.extend((Value(key), value))
+        super().__init__(*expressions)
+
+    def as_postgresql(self, compiler, connection, **extra_context):
+        return super().as_sql(compiler, connection, function='JSONB_BUILD_OBJECT', **extra_context)
+
+    def as_oracle(self, compiler, connection, **extra_context):
+        class ArgJoiner:
+            def join(self, args):
+                args = [' VALUE '.join(arg) for arg in zip(args[::2], args[1::2])]
+                return ', '.join(args)
+
+        clone = self.copy()
+        if clone.get_source_expressions():
+            clone.set_source_expressions([
+                Func(expression, function='TO_CHAR')
+                if expression.output_field.get_internal_type() == 'CharField'
+                else expression
+                for expression in clone.get_source_expressions()
+            ])
+            return super(JSONObject, clone).as_sql(compiler, connection, arg_joiner=ArgJoiner(), **extra_context)
+        clone.set_source_expressions([Cast(Value('{}'), output_field=JSONField())])
+        return super(JSONObject, clone).as_sql(compiler, connection, function='')
```

Comment:
`JSON_OBJECT()` on Oracle returns `VARCHAR2(4000)` by default so it will crash in many cases. It seems that there is too many changes between Oracle 12.2 and 18c to support both versions, also it's not very usable in Oracle 12.2 because it cannot return more than 4000 characters.

I think we should add `supports_json_object` feature flag and raise `NotSupportedError` on Oracle 12.2:
```python
raise NotSupportedError('JSONObject() requires Oracle 18+.')
```
we should also remove the current workaround with `TO_NCHAR()` and add [`RETURNING CLOB`](https://docs.oracle.com/en/database/oracle/oracle-database/18/sqlrf/JSON_OBJECT.html#GUID-1EF347AE-7FDA-4B41-AFE0-DD5A49E8B370) clause to support more than 4000 characters.

Reply:
I add `supports_json_object` which always equal `supports_json_field`. And need I test error "requires Oracle 18+"?

Reply:
I added a test.